R/simulate_data.R

```diff
@@ -0,0 +1,611 @@
+
+    rand_miss <- rbinom(n = pars_group$n*(n_visits-1), size = 1, prob = pars_group$prob_miss)
+    data$outcome[data$visit != "0"][rand_miss == 1] <- NA
+
+    return(data)
+}
+
+
+#' Simulate intercurrent event
+#'
+#' @param outcome Numeric variable that specifies the longitudinal outcome for a single group.
+#' @param visits Factor variable that specifies the visit of each assessment.
+#' @param ids Factor variable that specifies the id of each subject.
+#' @param prob_ice Numeric vector that specifies for each visit the probability of experiencing
+#' the ICE after the current visit for a subject with outcome equal to the mean at baseline.
+#' If a single numeric is provided, then the same probability is applied to each visit.
+#' @param or_outcome_ice Numeric value that specifies the odds ratio of the ICE corresponding to
+#' a +1 higher value of the outcome at the visit.
+#' @param baseline_mean Mean outcome value at baseline.
+#'
+#' @details The probability of the ICE after each visit is modeled according to the following
+#' logistic regression model:
+#' `~ 1 + I(visit == 0) + ... + I(visit == n_visits-1) + I((x-alpha))` where:
+#' - `n_visits` is the number of visits (including baseline).
+#' - `alpha` is the baseline outcome mean set via argument `baseline_mean`.
+#' The term `I((x-alpha))` specifies the dependency of the probability of the ICE on the current
+#' outcome value.
+#' The corresponding regression coefficients of the logistic model are defined as follows:
+#' The intercept is set to 0, the coefficients corresponding to discontinuation after each visit
+#' for a subject with outcome equal to
+#' the mean at baseline are set according to parameter `or_outcome_ice1`,
+#' and the regression coefficient associated with the covariate `I((x-alpha))` is set to
+#' `log(or_outcome_ice)`.
+#'
+#' @return A binary variable that takes value `1` if the corresponding outcome is affected
+#' by the ICE and `0` otherwise.
+#'
+#' @importFrom stats rbinom model.matrix binomial
+simulate_ice <- function(outcome, visits, ids, prob_ice, or_outcome_ice, baseline_mean) {
+
+    assert_that(
+        is.factor(visits) && length(visits) == length(outcome),
+        msg = "`visits` must be a factor of length equal to the length of `outcome`"
+    )
+
+    assert_that(
+        is.factor(ids) && length(ids) == length(outcome),
+        msg = "`ids` must be a factor of length equal to the length of `outcome`"
+    )
+
+    n_visits <- nlevels(visits)
+    if(length(prob_ice) == 1) {
+        prob_ice <- rep(prob_ice, n_visits)
+    }
+    prob_ice[prob_ice == 0] <- 1e-20
+    prob_ice[prob_ice == 1] <- 1-1e-15
+
+    create_model_ice <- function(n_visits, baseline_mean) {
+        model_ice <- "~ 1"
+        for(visit_num in seq.int(n_visits) - 1) {
+            model_ice <- paste0(model_ice, " + I(visit == ", visit_num,")")
+        }
+        model_ice <- paste0(model_ice, " + I((x - ", baseline_mean,"))")
```

Comment:
Why are you embedding the values into the formula environment instead of just putting them in the main dataframe ?